client/components/platform-checkout/save-user/checkout-page-save-user.js

```diff
@@ -14,22 +16,89 @@ import AdditionalInformation from './additional-information';
 import PhoneNumberInput from 'settings/phone-input';
 import Agreement from './agreement';
 import { getConfig } from 'utils/checkout';
+// eslint-disable-next-line import/no-unresolved
+import { extensionCartUpdate } from '@woocommerce/blocks-checkout';
 
-const CheckoutPageSaveUser = () => {
+const CheckoutPageSaveUser = ( { isBlocksCheckout } ) => {
 	const [ isSaveDetailsChecked, setIsSaveDetailsChecked ] = useState( false );
 	// eslint-disable-next-line no-unused-vars
 	const [ phoneNumber, setPhoneNumber ] = useState( '' );
 	const [ isPhoneValid, onPhoneValidationChange ] = useState( null );
+	const [ userDataSent, setUserDataSent ] = useState( false );
 	const isRegisteredUser = usePlatformCheckoutUser();
-	const {
-		isWCPayChosen,
-		isNewPaymentTokenChosen,
-	} = useSelectedPaymentMethod();
+	const { isWCPayChosen, isNewPaymentTokenChosen } = useSelectedPaymentMethod(
+		isBlocksCheckout
+	);
+	const cart = useDispatch( WC_STORE_CART );
+
+	const getPhoneFieldValue = () => {
+		let phoneFieldValue = '';
+		if ( isBlocksCheckout ) {
+			phoneFieldValue =
+				document.getElementById( 'phone' )?.value ||
+				document.getElementById( 'shipping-phone' )?.value ||
+				'';
+		} else {
+			// for classic checkout.
+			phoneFieldValue =
+				document.getElementById( 'billing_phone' )?.value || '';
+		}
+
+		return phoneFieldValue;
+	};
+
+	const sendExtensionData = useCallback(
+		( shouldClearData = false ) => {
+			const shippingPhone = document.getElementById( 'shipping-phone' )
+				?.value;
+			const billingPhone = document.getElementById( 'phone' )?.value;
+			const data = shouldClearData
+				? {}
+				: {
+						save_user_in_platform_checkout: isSaveDetailsChecked,
+						platform_checkout_user_phone_field: {
+							full: phoneNumber,
+						},
+				  };
+
+			extensionCartUpdate( {
+				namespace: 'platform-checkout',
+				data: data,
+			} ).then( () => {
+				setUserDataSent( ! shouldClearData );
+				// Cart returned from `extensionCartUpdate` clears these as these fields are not sent to backend by blocks when added.
+				// Setting them explicitly here to the previous user input.
+				cart.setShippingAddress( {
+					phone: shippingPhone,
+				} );
+				cart.setBillingAddress( {
+					phone: billingPhone,
+				} );
+			} );
+		},
+		[ isSaveDetailsChecked, phoneNumber, cart ]
+	);
+
+	const handleCheckboxClick = () => {
+		if ( isSaveDetailsChecked ) {
+			setPhoneNumber( null );
+			if ( isBlocksCheckout ) {
+				sendExtensionData( true );
+			}
+		} else {
+			setPhoneNumber( getPhoneFieldValue() );
+		}
+		setIsSaveDetailsChecked( ( v ) => ! v );
+	};
```

Comment:
I think this would be slightly easier to parse if we checked for the current value of the checkbox instead of the old value of `isSaveDetailsChecked`.

```javascript
const handleCheckboxClick = ( ev ) => {
	const isChecked = ev.target.checked;
	if ( isChecked ) {
		setPhoneNumber( getPhoneFieldValue() );
	} else {
		setPhoneNumber( null );
		if ( isBlocksCheckout ) {
			sendExtensionData( true );
		}
	}
	setIsSaveDetailsChecked( isChecked );
};
```

Feel free to ignore this comment 🙂.